core/src/main/scala/org/apache/spark/rdd/UnionRDD.scala

```diff
@@ -62,8 +62,14 @@ class UnionRDD[T: ClassTag](
     var rdds: Seq[RDD[T]])
   extends RDD[T](sc, Nil) {  // Nil since we implement getDependencies
 
+  // visible for testing
+  private[spark] val isPartitionEvalParallel: Boolean =
+    rdds.length > conf.getInt("spark.rdd.parallelListingThreshold", 10)
+
   override def getPartitions: Array[Partition] = {
-    val array = new Array[Partition](rdds.map(_.partitions.length).sum)
+    val parRDDs = if (isPartitionEvalParallel) rdds.par else rdds
```

Comment:
@rxin, does Spark have a default thread-pool for situations like this? I tend to agree with Sean's reasoning that we should use an existing thread-pool over the risk of contention with user code. Both concerns could be solved by having a pool for these one-off internal tasks.

Wouldn't user or third-party code blocking like you suggest slow down or stop the application anyway?